components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/ServiceDiscoverer.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.apimgt.core.impl;
+
+import org.wso2.carbon.apimgt.core.exception.ServiceDiscoveryException;
+import org.wso2.carbon.apimgt.core.models.Endpoint;
+import org.wso2.carbon.apimgt.core.util.APIMgtConstants.ServiceDiscoveryConstants;
+
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Abstract class to discover and list services in a cluster
+ */
+public abstract class ServiceDiscoverer {
+
+    private String namespaceFilter;
+    private HashMap<String, String> criteriaFilter;
+    int serviceEndpointIndex;
+
+    /**
+     * Initializes the parameters namespaceFilter, criteriaFilter,
+     * and calls #initImpl(java.util.HashMap) method
+     *
+     * @param implParameters  implementation parameters provided in the configuration
+     * @throws ServiceDiscoveryException if an error occurs in the implementation's initImpl method
+     */
+    public void init(HashMap<String, String> implParameters) throws ServiceDiscoveryException {
```

Comment:
Is there a specific reason to use use HashMap in the method signature? if not use Map instead of HashMap